### DISCUSSION TITLE
Should I use 'allowed' instead of 'parallelized' if my apply_ufunc are simply a variety of numpy functions?

### DISCUSSION THREAD

**Discussion by ahuang11** [pydata/xarray#4608] [Q&A] [answered]
""" http://xarray.pydata.org/en/stable/dask.html
For the majority of NumPy functions that are already wrapped by Dask, it’s usually a better idea to use the pre-existing dask.array function, by using either a pre-existing xarray methods or apply_ufunc() with dask='allowed'. Dask can often have a more efficient implementation that makes use of the specialized structure of a problem, unlike the generic speedups offered by dask='parallelized'.
""" 

If my function passed to xr.apply_ufunc only encompasses numpy functions (e.g. np.mean), but does not explicitly call dask.array functions (e.g. dask.array.mean), will the above still apply; i.e. should I use dask='allowed' rather than 'parallelized'?

Function similar to:
https://github.com/xarray-contrib/xskillscore/blob/master/xskillscore/core/np_deterministic.py#L527-L540

**Comment by dcherian** [comment, marked as answer]
Some of that advice needs to be updated. 

Things like `np.mean` will dispatch to `dask.array.mean` with recent enough versions of dask/numpy (see NEP-18). So `dask="allowed"` will work. This is easy to check: if the call to `apply_ufunc` ends up computing it means some function is not compatible with dask arrays. Either avoid that function or use `dask="parallelized"` if appropriate (the function can be applied independently on each block of the dask array, and does not need communication between blocks).
